### ISSUE TITLE
Request parameter with special character (plus) duplicated in request snippet

### ISSUE THREAD

**Issue by neubarth** [spring-projects/spring-restdocs#632]
When the value of a request parameter contains the plus sign (+), the parameter occurs twice in the generated snippet.
 
Example controller test:
```java
import static org.springframework.restdocs.mockmvc.MockMvcRestDocumentation.document;
import static org.springframework.restdocs.request.RequestDocumentation.parameterWithName;
import static org.springframework.restdocs.request.RequestDocumentation.requestParameters;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.restdocs.AutoConfigureRestDocs;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.restdocs.mockmvc.RestDocumentationResultHandler;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

import io.restassured.module.mockmvc.RestAssuredMockMvc;

@RunWith(SpringRunner.class)
@WebMvcTest(TestController.class)
@AutoConfigureRestDocs
public class TestControllerTest {

    @Autowired
    protected MockMvc mockMvc;

    @Before
    public void setUp() {
        RestAssuredMockMvc.mockMvc(mockMvc);
    }

    @Test
    public void test() {
        RestDocumentationResultHandler documentHandler = document(
                "{class-name}/{method-name}",
                requestParameters(parameterWithName("phone").description("phone number")));
        String phoneNumber = "+491234567";
        RestAssuredMockMvc.given().log().all().queryParam("phone", phoneNumber)
                .when().get("/test")
                .then().apply(documentHandler);
    }
}
```

Resulting snippet:
```
[source,http,options="nowrap"]
----
GET /test?phone=+491234567&phone=%2B491234567 HTTP/1.1
Host: localhost:8080
---
```

If i remove the + sign in `phoneNumber`, the snippet is correct, i.e. the parameter occurs only once.

Versions used:
spring-restdocs-mockmvc : 2.0.3
rest-assured : 4.0.0
Spring Boot 2.1.5

**Comment by marcgemis** [comment]
I have the same problem when the value is the empty string.

**Comment by wilkinsona** [comment]
@neubarth Thanks for the report. I've reproduced the problem.

@marcgemis That doesn't sound like the same problem to me. The problem reported here is due to the `+` being decoded into a ` ` and then encoded back into a `+` and being left as a `+` which results in it being encoded as `%2B`. As a result `phone` appears to have two values: `+491234567` and `%2B491234567`. I don't understand how that could occur with an empty string. Perhaps you could open a separate issue with a minimal sample the reproduces your problem?

**Comment by wilkinsona** [comment]
This is, at least in part, a bug in REST Assured. `+` is a reserved character in a query string and [RFC 2396](https://www.ietf.org/rfc/rfc2396.txt) says the following about escape encoding such characters:

> Normally, the only time escape encodings can safely be made is when the URI is being created from its component parts; each component may have its own set of characters that are reserved

It's REST Assured that's taking the `queryParam("phone", phoneNumber)` and adding it to the URL's query string so I think REST Assured should be encoding the `+` as `%2B` when it does so.

**Comment by wilkinsona** [comment]
The problem does not occur when REST Assured is used directly rather than via its MockMVC integration. When used directly it results in a `RequestSpecificationImpl` that returns `http://localhost:8080/?phone=%2B123456789` from `getURI()`. The problem also does not occur when using MockMvc directly.

I've tried to find a workaround, for example by using `.get("/test?phone=%2B491234567")` rather than `.queryParam("phone", phoneNumber)` to include the phone number query parameter, but this leads to double-encoding and the `%` being encoded.

@neubarth I think I've done all I can here. Please open a [REST Assured](https://github.com/rest-assured/rest-assured) issue and comment here with a link to it. I'm happy to help as needed and also to re-open this issue if my analysis turns out to be faulty.

Lastly, please note the REST Docs does not, yet, officially support REST Assured 4.0. I'd actually missed it being released so thanks for bringing that to my attention. I've opened https://github.com/spring-projects/spring-restdocs/issues/635.
